File: suites/07_cryptography_kernels/comparisons/crc32_checksum.c
#include <stdio.h>

static long long crc32_checksum(long long data_size) {
    long long crc = 4294967295LL;
    long long mask = 4294967295LL;
    long long poly = 3988292384LL;
    for (long long i = 0; i < data_size; ++i) {
        long long byte_val = (i * 1103515245LL + 12345LL) & 255;
        crc ^= byte_val;
        for (int bit = 0; bit < 8; ++bit) {
            if (crc & 1) crc = ((crc >> 1) & (mask >> 1)) ^ poly;
            else crc = (crc >> 1) & (mask >> 1);
        }
    }
    return (crc ^ mask) % 1000000007LL;
}

int main(void) { printf("%lld\n", crc32_checksum(16384)); return 0; }

File: suites/07_cryptography_kernels/comparisons/crc32_checksum.go
package main

import "fmt"

func crc32Checksum(dataSize int64) int64 {
	crc := int64(4294967295)
	mask := int64(4294967295)
	poly := int64(3988292384)
	for i := int64(0); i < dataSize; i++ {
		byteVal := (i*1103515245 + 12345) & 255
		crc ^= byteVal
		for bit := 0; bit < 8; bit++ {
			if (crc & 1) != 0 {
				crc = ((crc >> 1) & (mask >> 1)) ^ poly
			} else {
				crc = (crc >> 1) & (mask >> 1)
			}
		}
	}
	return (crc ^ mask) % 1000000007
}

func main() { fmt.Println(crc32Checksum(16384)) }
